task-impls/src/consensus.rs

```diff
@@ -248,29 +284,100 @@ where
 {
     async fn send_da(&self) {
         // TODO bf we need to send a DA proposal as soon as we are chosen as the leader
+        // ED: Added this in the DA task ^
     }
     async fn genesis_leaf(&self) -> Option<SequencingLeaf<TYPES>> {
         let consensus = self.consensus.read().await;
         let Some(genesis_view) = consensus.state_map.get(&TYPES::Time::genesis()) else {
-            warn!("Couldn't find genesis view in state map.");
+            error!("Couldn't find genesis view in state map.");
             return None;
         };
         let Some(leaf) = genesis_view.get_leaf_commitment() else {
-            warn!(
+            error!(
                 ?genesis_view,
                 "Genesis view points to a view without a leaf"
             );
             return None;
         };
         let Some(leaf) = consensus.saved_leaves.get(&leaf) else {
-            warn!("Failed to find genesis leaf.");
+            error!("Failed to find genesis leaf.");
             return None;
         };
         Some(leaf.clone())
     }
-    async fn vote_if_able(&self) {
+    async fn vote_if_able(&self) -> bool {
+        error!("In vote if able");
+
         if let Some(proposal) = &self.current_proposal {
-            if let Some(cert) = self.certs.get(&proposal.get_view_number()) {
+            // ED Need to account for the genesis DA cert
+            // // error!("in vote if able for proposal view {:?}", proposal.view_number);
+
+            if proposal.justify_qc.is_genesis() && proposal.view_number == ViewNumber::new(1) {
```

Comment:
I had kept this logic in the handling of the proposal so we didn't duplicate the vote token and proposal check when we get a DA cert since his is called in both places.

Reply:
I ended up running into an error where the `vote_if_able` function needs the leaf commitment to put in the vote.  In the future there is probably a cleaner way to handle this, but for now I duplicated some of the proposal checking code so I would have access to `leaf.commit()` inside `vote_if_able`  😄